File: android/app/src/main/kotlin/com/sopro/sopro/logging/Logger.kt
package com.sopro.sopro.logging

import android.util.Log
import java.time.Instant
import java.util.concurrent.CopyOnWriteArrayList

// Contrato de sink de saída. Recebe o LogEvent completo — cada sink extrai
// apenas os campos que precisa. Exceções lançadas de onEvent() são capturadas
// e logadas no Logcat; nunca interrompem o pipeline nem afetam os demais sinks.
fun interface LogSink {
    fun onEvent(event: LogEvent)
}

// Logger principal Kotlin — espelho do Logger Flutter.
//
// Todos os 6 métodos públicos (trace/debug/info/warn/error/fatal) delegam
// ao pipeline único emit(). Nenhuma lógica especial por nível fora de shouldEmit().
//
// Pipeline por evento:
//   1. shouldEmit() — filtra por nível
//   2. LogEvent criado com LogSanitizer aplicado ao payload
//   3. printToConsole() — LogCat (se enableConsole)
//   4. Dispatch para _sinks — independente de enableConsole
//
// Diferença vs Flutter: thread captura Thread.currentThread().name
// (Android é multi-thread; Dart é single-thread por isolate).
//
// Uso:
//   Logger.info("geofence_enter", feature = "geofence", payload = mapOf("id" to envId))
//   Logger.error("ble_gatt_error", exception = e, feature = "ble")
object Logger {

    // Lista de sinks registrados. Thread-safe: CopyOnWriteArrayList permite
    // iteração sem lock enquanto addSink() é raramente chamado.
    private val _sinks = CopyOnWriteArrayList<LogSink>()

    // Registra um sink adicional. Sem deduplicação automática — o chamador é
    // responsável por não registrar o mesmo sink duas vezes (use guard próprio).
    fun addSink(sink: LogSink) = _sinks.add(sink)

    private const val TAG = "Sopro"

    fun trace(
        message: String,
        payload: Map<String, Any?>? = null,
        feature: String? = null,
        action: String? = null,
        screen: String? = null,
        method: String? = null,
        correlationId: String? = null,
        exception: Throwable? = null,
        durationMs: Long? = null,
    ) = emit(LogLevel.TRACE, message, payload, feature, action, screen, method, correlationId, exception, durationMs)

    fun debug(
        message: String,
        payload: Map<String, Any?>? = null,
        feature: String? = null,
        action: String? = null,
        screen: String? = null,
        method: String? = null,
        correlationId: String? = null,
        exception: Throwable? = null,
        durationMs: Long? = null,
    ) = emit(LogLevel.DEBUG, message, payload, feature, action, screen, method, correlationId, exception, durationMs)

    fun info(
        message: String,
        payload: Map<String, Any?>? = null,
        feature: String? = null,
        action: String? = null,
        screen: String? = null,
        method: String? = null,
        correlationId: String? = null,
        exception: Throwable? = null,
        durationMs: Long? = null,
    ) = emit(LogLevel.INFO, message, payload, feature, action, screen, method, correlationId, exception, durationMs)

    fun warn(
        message: String,
        payload: Map<String, Any?>? = null,
        feature: String? = null,
        action: String? = null,
        screen: String? = null,
        method: String? = null,
        correlationId: String? = null,
        exception: Throwable? = null,
        durationMs: Long? = null,
    ) = emit(LogLevel.WARN, message, payload, feature, action, screen, method, correlationId, exception, durationMs)

    fun error(
        message: String,
        payload: Map<String, Any?>? = null,
        feature: String? = null,
        action: String? = null,
        screen: String? = null,
        method: String? = null,
        correlationId: String? = null,
        exception: Throwable? = null,
        durationMs: Long? = null,
    ) = emit(LogLevel.ERROR, message, payload, feature, action, screen, method, correlationId, exception, durationMs)

    fun fatal(
        message: String,
        payload: Map<String, Any?>? = null,
        feature: String? = null,
        action: String? = null,
        screen: String? = null,
        method: String? = null,
        correlationId: String? = null,
        exception: Throwable? = null,
        durationMs: Long? = null,
    ) = emit(LogLevel.FATAL, message, payload, feature, action, screen, method, correlationId, exception, durationMs)

    // ── Pipeline único ────────────────────────────────────────────────────────

    private fun emit(
        level: LogLevel,
        message: String,
        payload: Map<String, Any?>?,
        feature: String?,
        action: String?,
        screen: String?,
        method: String?,
        correlationId: String?,
        exception: Throwable?,
        durationMs: Long?,
    ) {
        if (!shouldEmit(level)) return

        val context = LogContext(
            deviceId = SessionManager.installationId,
            installationId = SessionManager.installationId,
            sessionId = SessionManager.sessionId,
            // correlationId explícito tem precedência sobre o CorrelationManager.
            correlationId = correlationId ?: CorrelationManager.currentCorrelationId,
            platform = LoggerConfiguration.platform,
            appVersion = LoggerConfiguration.appVersion,
            buildNumber = LoggerConfiguration.buildNumber,
            feature = feature,
            action = action,
            screen = screen,
            method = method,
            thread = Thread.currentThread().name,
            timestamp = Instant.now(),
        )

        val event = LogEvent(
            schemaVersion = LoggerConfiguration.SCHEMA_VERSION,
            level = level,
            message = message,
            context = context,
            payload = payload?.let { LogSanitizer.sanitize(it) },
            durationMs = durationMs,
            exception = exception,
            stackTraceString = exception?.stackTraceToString(),
        )

        if (LoggerConfiguration.enableConsole) {
            printToConsole(event)
        }

        // Despacha para todos os sinks em sequência. Falha de um sink é isolada:
        // não interrompe os demais. Console e sinks são independentes de enableConsole.
        for (sink in _sinks) {
            try {
                sink.onEvent(event)
            } catch (e: Exception) {
                Log.e(TAG, "[Logger] sink falhou — ${sink::class.simpleName}: ${e.message}")
            }
        }
    }

    // debugLogging == true → nível efetivo é TRACE (tudo passa).
    // debugLogging == false → respeita minimumLevel.
    private fun shouldEmit(level: LogLevel): Boolean {
        val effective = if (LoggerConfiguration.debugLogging) LogLevel.TRACE else LoggerConfiguration.minimumLevel
        return level.ordinal >= effective.ordinal
    }

    private fun printToConsole(event: LogEvent) {
        val androidPriority = when (event.level) {
            LogLevel.TRACE -> Log.VERBOSE
            LogLevel.DEBUG -> Log.DEBUG
            LogLevel.INFO -> Log.INFO
            LogLevel.WARN -> Log.WARN
            LogLevel.ERROR, LogLevel.FATAL -> Log.ERROR
        }
        val msg = if (LoggerConfiguration.enablePrettyPrint) {
            "[${event.level.label}] ${event.message} ${JsonFormatter.format(event)}"
        } else {
            "[${event.level.label}] ${event.message}"
        }
        Log.println(androidPriority, TAG, msg)
    }
}

File: lib/infrastructure/logging/core/logger.dart
import 'package:flutter/foundation.dart';

import 'correlation_manager.dart';
import 'json_log_formatter.dart';
import 'log_context.dart';
import 'log_event.dart';
import 'log_level.dart';
import 'log_sanitizer.dart';
import 'logger_configuration.dart';
import 'session_manager.dart';

// Logger principal do Sopro.
//
// Responsabilidades nesta fase:
//   - Criar LogEvent estruturado para cada chamada de log.
//   - Aplicar LogSanitizer sobre o payload.
//   - Formatar e emitir no console (se enableConsole == true).
//
// Fora do escopo desta fase (implementados em fases futuras):
//   - Fila de eventos (maxQueueSize / batchSize).
//   - Envio direto ao Supabase (permanece em AppLogger._send()).
//   - Retry automático (retryAttempts).
//
// Todos os métodos são estáticos para manter a mesma ergonomia do AppLogger
// legado e evitar injeção de dependência onde não é necessária.
class Logger {
  Logger._();

  // Nível mais granular — apenas para rastreamento interno de fluxo.
  static void trace(
    String message, {
    Map<String, dynamic>? payload,
    String? feature,
    String? action,
    String? screen,
    String? method,
    Object? exception,
    StackTrace? stackTrace,
    int? durationMs,
  }) =>
      _emit(
        LogLevel.trace,
        message,
        payload: payload,
        feature: feature,
        action: action,
        screen: screen,
        method: method,
        exception: exception,
        stackTrace: stackTrace,
        durationMs: durationMs,
      );

  // Informações de diagnóstico úteis durante desenvolvimento.
  static void debug(
    String message, {
    Map<String, dynamic>? payload,
    String? feature,
    String? action,
    String? screen,
    String? method,
    Object? exception,
    StackTrace? stackTrace,
    int? durationMs,
  }) =>
      _emit(
        LogLevel.debug,
        message,
        payload: payload,
        feature: feature,
        action: action,
        screen: screen,
        method: method,
        exception: exception,
        stackTrace: stackTrace,
        durationMs: durationMs,
      );

  // Eventos operacionais normais (app_start, geofence_enter, trigger_fired…).
  static void info(
    String message, {
    Map<String, dynamic>? payload,
    String? feature,
    String? action,
    String? screen,
    String? method,
    Object? exception,
    StackTrace? stackTrace,
    int? durationMs,
  }) =>
      _emit(
        LogLevel.info,
        message,
        payload: payload,
        feature: feature,
        action: action,
        screen: screen,
        method: method,
        exception: exception,
        stackTrace: stackTrace,
        durationMs: durationMs,
      );

  // Situações inesperadas que não interrompem o fluxo mas merecem atenção.
  static void warn(
    String message, {
    Map<String, dynamic>? payload,
    String? feature,
    String? action,
    String? screen,
    String? method,
    Object? exception,
    StackTrace? stackTrace,
    int? durationMs,
  }) =>
      _emit(
        LogLevel.warn,
        message,
        payload: payload,
        feature: feature,
        action: action,
        screen: screen,
        method: method,
        exception: exception,
        stackTrace: stackTrace,
        durationMs: durationMs,
      );

  // Falhas recuperáveis (ble_error, photon_http_error, notification_error…).
  static void error(
    String message, {
    Map<String, dynamic>? payload,
    String? feature,
    String? action,
    String? screen,
    String? method,
    Object? exception,
    StackTrace? stackTrace,
    int? durationMs,
  }) =>
      _emit(
        LogLevel.error,
        message,
        payload: payload,
        feature: feature,
        action: action,
        screen: screen,
        method: method,
        exception: exception,
        stackTrace: stackTrace,
        durationMs: durationMs,
      );

  // Falhas críticas que comprometem a operação do app.
  static void fatal(
    String message, {
    Map<String, dynamic>? payload,
    String? feature,
    String? action,
    String? screen,
    String? method,
    Object? exception,
    StackTrace? stackTrace,
    int? durationMs,
  }) =>
      _emit(
        LogLevel.fatal,
        message,
        payload: payload,
        feature: feature,
        action: action,
        screen: screen,
        method: method,
        exception: exception,
        stackTrace: stackTrace,
        durationMs: durationMs,
      );

  // ─── Implementação interna ────────────────────────────────────────────────

  static void _emit(
    LogLevel level,
    String message, {
    Map<String, dynamic>? payload,
    String? feature,
    String? action,
    String? screen,
    String? method,
    Object? exception,
    StackTrace? stackTrace,
    int? durationMs,
  }) {
    if (!_shouldEmit(level)) return;

    final sanitized =
        payload != null ? LogSanitizer.sanitize(payload) : null;

    final event = LogEvent(
      level: level,
      message: message,
      context: LogContext(
        deviceId: SessionManager.installationId,
        installationId: SessionManager.installationId,
        sessionId: SessionManager.sessionId,
        correlationId: CorrelationManager.currentCorrelationId,
        platform: LoggerConfiguration.platform,
        appVersion: LoggerConfiguration.appVersion,
        buildNumber: LoggerConfiguration.buildNumber,
        feature: feature,
        action: action,
        screen: screen,
        method: method,
        thread: null,
        timestamp: DateTime.now().toUtc(),
      ),
      payload: sanitized,
      durationMs: durationMs,
      exception: exception,
      stackTrace: stackTrace,
    );

    if (LoggerConfiguration.enableConsole) {
      _printToConsole(event);
    }
  }

  // trace e debug são suprimidos fora do modo debug para reduzir ruído.
  static bool _shouldEmit(LogLevel level) {
    if (!LoggerConfiguration.debugLogging &&
        (level == LogLevel.trace || level == LogLevel.debug)) {
      return false;
    }
    return true;
  }

  static void _printToConsole(LogEvent event) {
    if (LoggerConfiguration.enablePrettyPrint) {
      final json = JsonLogFormatter.format(event);
      debugPrint('[Sopro][${event.level.label}] ${event.message} $json');
    } else {
      debugPrint('[Sopro][${event.level.label}] ${event.message}');
    }
  }
}
